lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/f2fCri/ContractTest.java

```diff
@@ -0,0 +1,411 @@
+        // painful.
+        when(mockSigner.sign(any(), any()))
+                .thenReturn(
+                        new Base64URL(
+                                "hXYrKJ_W9YItUbZxu3T63gQgScVoSMqHZ43UPfdB8im8L4d0mZPLC6BlwMJSsfjiAyU1y3c37vm-rV8kZo2uyw"));
+        when(mockSigner.supportedJWSAlgorithms()).thenReturn(Set.of(JWSAlgorithm.ES256));
+        when(mockSecureTokenHelper.generate())
+                .thenReturn("ScnF4dGXthZYXS_5k85ObEoSU04W-H3qa_p6npv2ZUY");
+
+        // We need to generate a fixed request, so we set the secure token and expiry to constant
+        // values.
+        var underTest =
+                new CriApiService(
+                        mockConfigService, mockSigner, mockSecureTokenHelper, CURRENT_TIME);
+
+        CriApiException exception =
+                assertThrows(
+                        CriApiException.class,
+                        () -> {
+                            underTest.fetchAccessToken(
+                                    new CriCallbackRequest(
+                                            "dummyAuthCode",
+                                            credentialIssuerConfig.getClientId(),
+                                            "dummySessionId",
+                                            null,
+                                            "dummyState",
+                                            null,
+                                            null,
+                                            "dummyIpAddress",
+                                            "dummyFeatureSet"),
+                                    new CriOAuthSessionItem(
+                                            "dummySessionId",
+                                            "dummyOAuthSessionId",
+                                            "dummyCriId",
+                                            "dummyConnection",
+                                            900));
+                        });
+
+        // Assert
+        assertThat(exception.getErrorResponse(), is(ErrorResponse.INVALID_TOKEN_REQUEST));
+        assertThat(exception.getHttpStatusCode(), is(HTTPResponse.SC_BAD_REQUEST));
+    }
+
+    @NotNull
+    private static CredentialIssuerConfig getMockF2FCredentialIssuerConfig(
+            MockServer mockServer, URI clientCallbackUrl) throws URISyntaxException {
+        return new CredentialIssuerConfig(
+                new URI("http://localhost:" + mockServer.getPort() + "/token"),
+                new URI("http://localhost:" + mockServer.getPort() + "/credential"),
+                new URI("http://localhost:" + mockServer.getPort() + "/authorize"),
+                IPV_CORE_CLIENT_ID,
+                CRI_SIGNING_PRIVATE_KEY_JWK,
+                CRI_RSA_ENCRYPTION_PUBLIC_JWK,
+                "dummyF2fComponentId",
+                clientCallbackUrl,
+                true,
+                false);
+    }
+}
```

Comment:
I did some refactoring of the driving licence test file to e.g. move the pacts to directly above the relevant test (instead of all the pacts then all the tests) and pull some code into helpers. Might be worth taking a look at that and seeing what you think about applying some or all of it here.